File: 0053-maximum-subarray/0053-maximum-subarray.py
"""
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        count = 0
        max_sum = nums[0]
        for i in range(len(nums)):
            if count + nums[i] < 0:
                if nums[i] > max_sum:
                    max_sum = nums[i]
                count = 0
            else:
                count += nums[i]
                if count > max_sum:
                    max_sum = count
        
        return max_sum
"""

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        sums = [0] * len(nums)
        sums[0] = nums[0]
        
        for i in range(1, len(nums)):
            sums[i] = max(sums[i-1] + nums[i], nums[i])
        
        return max(sums)

File: 0053-maximum-subarray/0053-maximum-subarray.go
func maxSubArray(nums []int) int {
    maxSum, curSum := nums[0], nums[0]
    
    for _, n := range nums[1:] {
        curSum = max(curSum + n, n) 
        maxSum = max(maxSum, curSum)
    }
    
    return maxSum
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}